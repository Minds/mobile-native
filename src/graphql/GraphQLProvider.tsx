import { PropsWithChildren } from 'react';
import {
  ApolloClient,
  ApolloLink,
  ApolloProvider,
  HttpLink,
  InMemoryCache,
} from '@apollo/client';
import { GRAPHQL_CONFIG, USE_APOLLO } from '~/config/Config';

const client = new ApolloClient({
  link: ApolloLink.split(
    operation => operation.getContext().clientName === 'strapi',
    new HttpLink({
      ...GRAPHQL_CONFIG.strapi,
    }),
    new HttpLink({
      ...GRAPHQL_CONFIG.minds,
    }),
  ),
  cache: new InMemoryCache(),
});

export const GraphQLProvider = ({ children }: PropsWithChildren) =>
  USE_APOLLO ? (
    <ApolloProvider client={client}>{children}</ApolloProvider>
  ) : (
    <>{children}</>
  );
