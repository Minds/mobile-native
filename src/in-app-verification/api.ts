import apiService from '~/common/services/api.service';
import { IS_IOS } from '~/config/Config';

export const api = {
  requestCode(deviceId: string, deviceToken: string) {
    return apiService.post(`/api/v3/verification/${deviceId}`, {
      device_type: IS_IOS ? 'ios' : 'android',
      device_token: deviceToken,
    });
  },
  submitVerification(
    deviceId: string,
    image: any,
    video: any,
    sensorData: string,
    location: string,
  ) {
    return apiService.upload(
      `api/v3/verification/${deviceId}/verify`,
      { image, video },
      {
        sensor_data: sensorData,
        device_type: IS_IOS ? 'ios' : 'android',
        geo: location,
      },
    );
  },
};
