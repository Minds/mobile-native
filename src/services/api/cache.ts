abstract class BaseCache {
  private cache: Record<string, CacheValue> = {};
  get = (key: string) => this.cache[key];
  set = (key: string, value: CacheValue) => (this.cache[key] = value);
  flushAll = () => (this.cache = {});
}

class DefaultCache extends BaseCache {}

interface CacheValue {
  etag: string;
  value: any;
}

export class Cache {
  static instance: Cache;
  cache: BaseCache;

  static getInstance = () => {
    if (!this.instance) {
      this.instance = new Cache(new DefaultCache());
    }
    return this.instance;
  };

  static get = (uuid: string) => this.getInstance().cache.get(uuid);

  static set = (uuid: string, etag: string, value: any) =>
    this.getInstance().cache.set(uuid, { etag, value });

  static reset = () => this.getInstance().cache.flushAll();

  constructor(cache: BaseCache) {
    this.cache = cache;
  }
}

const byLowerCase = (toFind: string) => (value: string) =>
  value.toLowerCase() === toFind;

export const getHeaderCaseInsensitive = (
  headerName: string,
  headers: Record<string, string> = {},
) => {
  const key = Object.keys(headers).find(byLowerCase(headerName));
  return key ? headers[key] : undefined;
};
