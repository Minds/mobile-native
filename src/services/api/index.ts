export { default as ApiService } from './api.service';
