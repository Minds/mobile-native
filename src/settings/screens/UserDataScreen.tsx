import React from 'react';

import { CheckBox } from 'react-native-elements';
import { observer } from 'mobx-react';
import { View } from 'react-native';

import MText from '~/common/components/MText';
import { Screen } from '~/common/ui';
import { showNotification } from 'AppMessages';
import { Button } from '~/common/ui/buttons';
import { useDeletePostHogPersonMutation } from '~/graphql/api';

import sp from '~/services/serviceProvider';

const UserDataScreen = observer(() => {
  const theme = sp.styles.style;

  const deleteDataMutation = useDeletePostHogPersonMutation();

  const isOptOut = () => sp.config.getSettings()?.posthog?.opt_out;

  const toggleOptOut = () => {
    sp.resolve('analytics').setOptOut(!isOptOut());

    try {
      sp.resolve('settings').submitSettings({
        opt_out_analytics: isOptOut(),
      });
      showNotification(sp.i18n.t('settings.saved'), 'info');
    } catch (err) {
      showNotification(sp.i18n.t('errorMessage'));
      console.error(err);
    }
  };

  const deleteData = async () => {
    try {
      await deleteDataMutation.mutateAsync({});
      showNotification(
        sp.i18n.t('settings.userData.deletedSuccess'),
        'success',
      );
    } catch (err) {
      showNotification(sp.i18n.t('settings.userData.deletedError'), 'warning');
      console.log(err);
    }
  };

  return (
    <Screen scroll>
      <MText
        style={[
          theme.paddingHorizontal4x,
          theme.paddingTop4x,
          theme.colorSecondaryText,
        ]}>
        Your user data consists of two primary categories of data:
      </MText>
      <MText
        style={[
          theme.paddingHorizontal4x,
          theme.paddingTop4x,
          theme.colorSecondaryText,
        ]}>
        (1) A record of your public contributions to the network, including your
        public profile, posts, media uploads, and votes. These are stored
        indefinitely (or until you remove them) so that they are available to
        you and other public users of the app.
      </MText>
      <MText
        style={[
          theme.paddingHorizontal4x,
          theme.paddingTop4x,
          theme.colorSecondaryText,
        ]}>
        (2) Observational analytics, including which pages and posts you've
        seen, which buttons you've clicked on, and contextual information like
        which browser you're using, and where you came from. Observational
        analytics are used internally and allow us to reduce spam, diagnose
        problems with the app, and understand how to improve the service. They
        also allow us to enhance your experience, e.g. by recommending only
        posts that you haven't already seen before.
      </MText>
      <MText
        style={[
          theme.paddingHorizontal4x,
          theme.paddingTop4x,
          theme.colorSecondaryText,
        ]}>
        Observational analytics are not shared with outside companies for
        tracking purposes. Moreover, we offer you the transparency to download
        your personal analytics (coming soon) to see what's collected, and
        optionally opt-out of collection if you prefer.
      </MText>

      <MText
        style={[
          theme.paddingHorizontal4x,
          theme.paddingTop4x,
          theme.colorSecondaryText,
        ]}>
        <MText
          style={[
            theme.paddingHorizontal4x,
            theme.paddingTop4x,
            theme.colorAlert,
          ]}>
          Note:&nbsp;
        </MText>
        By opting out of observational analytics, you may see recommendations
        for posts you've already seen. And because observational analytics allow
        us to detect spam engagements, opting out means that your app
        engagements will not contribute to trending rankings and engagement
        rewards.
      </MText>

      <CheckBox
        title={sp.i18n.t('settings.userData.optOut')}
        containerStyle={[theme.checkbox, theme.marginLeft4x]}
        checked={isOptOut()}
        onPress={toggleOptOut}
        textStyle={theme.colorPrimaryText}
      />

      <MText
        style={[
          theme.fontXL,
          theme.paddingHorizontal4x,
          theme.paddingTop4x,
          theme.colorSecondaryText,
        ]}>
        User data transparency
      </MText>
      <MText
        style={[
          theme.paddingHorizontal4x,
          theme.paddingTop4x,
          theme.paddingBottom4x,
          theme.colorSecondaryText,
        ]}>
        It's important to us that you're in control of your user data. If you
        wish to delete the observational data, you may do so below.
      </MText>

      <View
        style={[
          theme.paddingHorizontal4x,
          theme.paddingTop4x,
          theme.paddingBottom8x,
        ]}>
        <Button
          top="XXL"
          onPress={deleteData}
          type="warning"
          testID="delete-data-btn">
          {sp.i18n.t('settings.userData.deleteData')}
        </Button>
      </View>
    </Screen>
  );
});

export default UserDataScreen;
