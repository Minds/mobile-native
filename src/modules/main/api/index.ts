import { ApiConnector, to } from 'services/api';

// import { getSubscriptionsAndBalances } from 'app/sdk/subscriptions';
export async function login({
  username,
  passcode,
}: {
  username?: string;
  passcode?: string;
}): Promise<unknown> {
  const [response] = await to(
    ApiConnector.getInstance()
      .post('/v1/login', {
        username,
        passcode,
        scope: 'subscription:read',
      })
      .then(result => result.data),
  );
  return response;
}

export async function registeredDevice({
  username,
  passcode,
}: {
  username?: string;
  passcode?: string;
}): Promise<boolean> {
  const response = await ApiConnector.getInstance()
    .post('/v1/verifications/registeredDevice', {
      username,
      passcode,
    })
    .then(resp => resp.status === 200)
    .catch(() => false);
  return !!response;
}

export async function requestOtpForDeviceChange(): Promise<unknown> {
  const [response] = await to(
    ApiConnector.getInstance()
      .post('/v1/devices/verification', {})
      .then(result => result.data),
  );
  return response;
}

export async function verifyOtpForDeviceChange({
  verificationId,
  otp,
}: {
  verificationId?: string;
  otp?: string;
}): Promise<unknown> {
  const [response] = await to(
    ApiConnector.getInstance()
      .post(`/v1/devices/verification/${verificationId}`, { otp })
      .then(result => result.data),
  );
  return response;
}
// end-of-api-methods - HYGEN
