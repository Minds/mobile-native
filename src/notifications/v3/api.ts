import UserModel from '~/channel/UserModel';
import { fetchFeedPage, PageResponse } from '~/services';
import NotificationModel from './notification/NotificationModel';

const mapToModel = (notification: NotificationModel) => {
  notification = NotificationModel.create(notification);
  if (notification.merged_from && notification.merged_from.length > 0) {
    notification.merged_from = UserModel.createMany(notification.merged_from);
  }
  return notification;
};

/**
 * Fetch a page from the notifications feed
 * @param offset string
 * @param filter string
 */
export const fetchNotificationsPage = async (
  offset = '',
  filter = '',
): Promise<PageResponse<NotificationModel>> => {
  const params = {
    filter,
    limit: 15,
    offset,
  };

  console.log('FETCHING NOTIFICATIONS', offset);

  const data = await fetchFeedPage<NotificationModel>(
    'api/v3/notifications/list',
    'notifications',
    params,
    mapToModel,
  );

  return data;
};
