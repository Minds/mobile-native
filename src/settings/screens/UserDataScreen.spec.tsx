import * as React from 'react';
import { fireEvent, render, screen } from '@testing-library/react-native';
import UserDataScreen from './UserDataScreen';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import analyticsService from '~/common/services/analytics.service';
import settingsService from '../SettingsApiService';
import mindsConfigService from '~/common/services/minds-config.service';
import { useDeletePostHogPersonMutation } from '~/graphql/api';

jest.mock('~/common/services/analytics.service', () => ({
  setOptOut: jest.fn(),
}));
jest.mock('../SettingsService', () => ({
  submitSettings: jest.fn(),
}));

jest.mock('~/graphql/api');

const useDeletePostHogPersonMutationMocked =
  useDeletePostHogPersonMutation as jest.MockedFn<
    typeof useDeletePostHogPersonMutation
  >;

describe('UserDataScreen', () => {
  let queryClient: QueryClient, wrapper;

  beforeEach(() => {
    queryClient = new QueryClient();
    wrapper = ({ children }) => (
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    );
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should load data and render correctly', async () => {
    render(<UserDataScreen />, { wrapper });

    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('should set opt out to true when clicking on unchecked checkbox', () => {
    render(<UserDataScreen />, { wrapper });

    jest.spyOn(analyticsService, 'setOptOut');
    jest.spyOn(settingsService, 'submitSettings');

    const checkbox = screen.getByTestId('checkbox');
    fireEvent.press(checkbox);

    expect(analyticsService.setOptOut).toHaveBeenCalledTimes(1);
    expect(analyticsService.setOptOut).toHaveBeenCalledWith(true);
    expect(settingsService.submitSettings).toHaveBeenCalled();
  });

  it('should set opt out to true when clicking on unchecked checkbox', () => {
    jest.spyOn(mindsConfigService, 'getSettings').mockReturnValue({
      posthog: {
        opt_out: true,
      },
    });

    jest.spyOn(analyticsService, 'setOptOut');
    jest.spyOn(settingsService, 'submitSettings');

    render(<UserDataScreen />, { wrapper });

    const checkbox = screen.getByTestId('checkbox');
    fireEvent.press(checkbox);

    expect(analyticsService.setOptOut).toHaveBeenCalledTimes(1);
    expect(analyticsService.setOptOut).toHaveBeenCalledWith(false);
    expect(settingsService.submitSettings).toHaveBeenCalled();
  });

  it('should delete data when clicking button', async () => {
    const mockMutate = jest.fn();

    useDeletePostHogPersonMutationMocked.mockReturnValue({
      mutateAsync: mockMutate,
    } as any);

    render(<UserDataScreen />, { wrapper });

    const btn = screen.getByTestId('delete-data-btn');
    fireEvent.press(btn);

    expect(mockMutate).toHaveBeenCalled();
  });
});
