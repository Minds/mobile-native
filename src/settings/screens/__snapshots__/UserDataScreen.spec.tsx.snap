// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDataScreen should load data and render correctly 1`] = `
<View
  style={
    [
      [
        {
          "flex": 1,
        },
        {
          "backgroundColor": "#010100",
        },
      ],
      [
        {
          "alignSelf": "center",
          "maxWidth": 750,
          "width": "100%",
        },
      ],
    ]
  }
>
  <RCTScrollView
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    scrollEnabled={false}
    showsVerticalScrollIndicator={false}
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <Text
        style={
          [
            {
              "color": "#F2F2F2",
            },
            {
              "fontFamily": "Roboto_400Regular",
            },
            [
              {
                "paddingHorizontal": 16,
              },
              {
                "paddingTop": 16,
              },
              {
                "color": "#b9bbc0",
              },
            ],
          ]
        }
      >
        Your user data consists of two primary categories of data:
      </Text>
      <Text
        style={
          [
            {
              "color": "#F2F2F2",
            },
            {
              "fontFamily": "Roboto_400Regular",
            },
            [
              {
                "paddingHorizontal": 16,
              },
              {
                "paddingTop": 16,
              },
              {
                "color": "#b9bbc0",
              },
            ],
          ]
        }
      >
        (1) A record of your public contributions to the network, including your public profile, posts, media uploads, and votes. These are stored indefinitely (or until you remove them) so that they are available to you and other public users of the app.
      </Text>
      <Text
        style={
          [
            {
              "color": "#F2F2F2",
            },
            {
              "fontFamily": "Roboto_400Regular",
            },
            [
              {
                "paddingHorizontal": 16,
              },
              {
                "paddingTop": 16,
              },
              {
                "color": "#b9bbc0",
              },
            ],
          ]
        }
      >
        (2) Observational analytics, including which pages and posts you've seen, which buttons you've clicked on, and contextual information like which browser you're using, and where you came from. Observational analytics are used internally and allow us to reduce spam, diagnose problems with the app, and understand how to improve the service. They also allow us to enhance your experience, e.g. by recommending only posts that you haven't already seen before.
      </Text>
      <Text
        style={
          [
            {
              "color": "#F2F2F2",
            },
            {
              "fontFamily": "Roboto_400Regular",
            },
            [
              {
                "paddingHorizontal": 16,
              },
              {
                "paddingTop": 16,
              },
              {
                "color": "#b9bbc0",
              },
            ],
          ]
        }
      >
        Observational analytics are not shared with outside companies for tracking purposes. Moreover, we offer you the transparency to download your personal analytics (coming soon) to see what's collected, and optionally opt-out of collection if you prefer.
      </Text>
      <Text
        style={
          [
            {
              "color": "#F2F2F2",
            },
            {
              "fontFamily": "Roboto_400Regular",
            },
            [
              {
                "paddingHorizontal": 16,
              },
              {
                "paddingTop": 16,
              },
              {
                "color": "#b9bbc0",
              },
            ],
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#F2F2F2",
              },
              {
                "fontFamily": "Roboto_400Regular",
              },
              [
                {
                  "paddingHorizontal": 16,
                },
                {
                  "paddingTop": 16,
                },
                {
                  "color": "#fd8b7c",
                },
              ],
            ]
          }
        >
          Note:Â 
        </Text>
        By opting out of observational analytics, you may see recommendations for posts you've already seen. And because observational analytics allow us to detect spam engagements, opting out means that your app engagements will not contribute to trending rankings and engagement rewards.
      </Text>
      <View
        accessibilityRole="checkbox"
        accessibilityState={
          {
            "checked": false,
          }
        }
        checkedIcon="check-square-o"
        onPress={[Function]}
        size={24}
        style={
          {
            "backgroundColor": "transparent",
            "borderColor": "#ededed",
            "borderRadius": 3,
            "borderWidth": 0,
            "margin": 5,
            "marginLeft": 16,
            "marginRight": 10,
            "padding": 10,
            "paddingLeft": 0,
          }
        }
        testID="checkbox"
        uncheckedColor="#bfbfbf"
        uncheckedIcon="square-o"
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          FontAwesome
          <Text
            style={
              {
                "color": "#F2F2F2",
                "fontWeight": "bold",
                "marginLeft": 10,
                "marginRight": 10,
              }
            }
            testID="checkboxTitle"
          >
            Opt-out of observational analtytics
          </Text>
        </View>
      </View>
      <Text
        style={
          [
            {
              "color": "#F2F2F2",
            },
            {
              "fontFamily": "Roboto_400Regular",
            },
            [
              {
                "fontSize": 18,
              },
              {
                "paddingHorizontal": 16,
              },
              {
                "paddingTop": 16,
              },
              {
                "color": "#b9bbc0",
              },
            ],
          ]
        }
      >
        User data transparency
      </Text>
      <Text
        style={
          [
            {
              "color": "#F2F2F2",
            },
            {
              "fontFamily": "Roboto_400Regular",
            },
            [
              {
                "paddingHorizontal": 16,
              },
              {
                "paddingTop": 16,
              },
              {
                "paddingBottom": 16,
              },
              {
                "color": "#b9bbc0",
              },
            ],
          ]
        }
      >
        It's important to us that you're in control of your user data. If you wish to delete the observational data, you may do so below.
      </Text>
      <View
        style={
          [
            {
              "paddingHorizontal": 16,
            },
            {
              "paddingTop": 16,
            },
            {
              "paddingBottom": 32,
            },
          ]
        }
      >
        <View
          style={
            {
              "alignSelf": "stretch",
              "marginTop": 24,
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignSelf": "stretch",
              }
            }
            testID="delete-data-btn"
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "borderRadius": 100,
                    "flexDirection": "column",
                    "justifyContent": "center",
                    "overflow": "hidden",
                  },
                  undefined,
                  {
                    "minHeight": 42,
                    "paddingHorizontal": 24,
                    "paddingVertical": 2,
                  },
                  undefined,
                  undefined,
                  undefined,
                  false,
                ]
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#fd8b7c",
                    "borderColor": "transparent",
                    "borderRadius": 100,
                    "borderWidth": 1,
                    "bottom": 0,
                    "justifyContent": "center",
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
              />
              <View
                collapsable={false}
                style={
                  {
                    "backgroundColor": "rgba(0, 0, 0, 0.1)",
                    "bottom": 0,
                    "left": 0,
                    "opacity": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
              />
              <View
                collapsable={false}
                style={
                  {
                    "transform": [
                      {
                        "scale": 1,
                      },
                    ],
                  }
                }
              >
                <Text
                  numberOfLines={1}
                  style={
                    [
                      {
                        "color": "#FFFFFF",
                      },
                      {
                        "fontFamily": "Roboto_500Medium",
                      },
                      {
                        "fontSize": 18,
                        "lineHeight": 24,
                      },
                      {
                        "textAlign": "left",
                      },
                    ]
                  }
                >
                  Delete my observational data
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
